### PR TITLE
Add public API

### PR DESCRIPTION
**Pull request by silesky**
1. add criticalTasks to register/deregister, so registration only happens after plugins are loaded2
2. tweak EventEmitter interface (see README.md)
3. return void

### CODE REVIEW

File: packages/core/src/emitter/index.ts [hunk 1]
```diff
@@ -1,29 +1,61 @@
...
+ * Event Emitter that takes the expected contract as a generic
+ * @example
+ * ```ts
+ *  type Contract = {
+ *    delivery_success: [DeliverySuccessResponse, Metrics],
+ *    delivery_failure: [DeliveryError]
+ * }
+ *  new Emitter<Contract>()
+ *  .on('delivery_success', (res, metrics) => ...)
+ *  .on('delivery_failure', (err) => ...)
```
Comment (chrisradek) [inline packages/core/src/emitter/index.ts:15]: Love that you added an example here - it gets hard to parse the types sometimes :smile:

File: packages/core/src/events/interfaces.ts [hunk 2]
```diff
@@ -26,7 +26,7 @@ export type Integrations = {
 }
 
 // renamed
-export type CoreOptions = {
+export interface CoreOptions {
```
Comment (chrisradek) [inline packages/core/src/events/interfaces.ts:29]: Just curious - was this change needed or just being consistent?

File: packages/node/README.md [hunk 3]
```diff
@@ -5,22 +5,25 @@ https://segment.com/docs/connections/sources/catalog/libraries/server/node/
...
 
-
+## Basic Usage
 ```ts
-// TODO: finalize API
-import { load } from '@segment/analytics-node'
+// analytics.ts
+import { AnalyticsNode } from '@segment/analytics-node'
 
+export const analytics = new AnalyticsNode({ writeKey: '<MY_WRITE_KEY>' })
```
Comment (chrisradek) [inline packages/node/README.md:13]: 🎉

File: packages/node/README.md [hunk 4]
```diff
@@ -5,22 +5,25 @@ https://segment.com/docs/connections/sources/catalog/libraries/server/node/
...
-router.post('/user',  async (req, res) => {
-  ajsPromise.then((ajs) => ajs.track(user.id, "some event"))
-  res.json(user)
-})
+analytics.identify('Test User', { loggedIn: true }, { userId: "123456" })
+analytics.track('hello world', {}, { userId: "123456" })
 
+```
 
+# Event Emitter (Advanced Usage)
```
Comment (chrisradek) [inline packages/node/README.md:24]: More of a todo, but we should make sure to document how to handle errors. (maybe that won't be apparent to us until the plugin is added)

**Review by chrisradek — APPROVED** [review]
I think this looks great!

File: packages/core/src/events/interfaces.ts (hunk 2 above)
Comment (silesky) [reply packages/core/src/events/interfaces.ts:29]: You mean the rename? I thought it made sense to prefix the things that were going to become base classes / interface (e.g. in node, we extend these options with Identity). However, being totally consistent here is hard.

File: packages/core/src/emitter/index.ts (hunk 1 above)
Comment (silesky) [reply packages/core/src/emitter/index.ts:15]: Thanks!

File: packages/node/README.md (hunk 4 above)
Comment (silesky) [reply packages/node/README.md:24]: I created a new `dispatchWithErrorHandling` function that is worth a look -- it emits an 'error' whenever there's a delivery failure due to for example, a validation failure, or an outright rejection (such as a network error)

File: packages/core/src/analytics/dispatch.ts [hunk 5]
```diff
@@ -57,3 +57,21 @@ export async function dispatch(
...
+  try {
+    const ctx = await dispatch(event, queue, emitter, options)
+    if (ctx.failedDelivery()) {
+      throw ctx
+    } else {
+      emitter.emit(event.type, ctx)
+      return ctx
+    }
+  } catch (ctx) {
+    emitter.emit('error', ctx)
```
Comment (silesky) [inline packages/core/src/analytics/dispatch.ts:74]: do we want to wrap all our errors in custom objects to give more metadata about their context -- some errors may not have actual context object.

File: packages/core/src/events/index.ts [hunk 6]
```diff
@@ -28,7 +28,7 @@ export class EventFactory {
       ...this.baseEvent(),
       event,
       type: 'track',
-      properties,
+      properties: properties ?? {},
```
Comment (pooyaj) [inline packages/core/src/events/index.ts:31]: Would this be a breaking change in the eventual `track` playload structure (i.e., after this change track payload will always have a `properties` object regardless )?

File: packages/node/src/app/analytics-node.ts [hunk 7]
```diff
@@ -39,294 +57,263 @@ export interface AnalyticsSettings {
   writeKey: string
   timeout?: number
   plugins?: CorePlugin[]
+  integrations?: Integrations
+  retryQueue?: boolean
```
Comment (pooyaj) [inline packages/node/src/app/analytics-node.ts:61]: 🤔 I think eventually we can get rid of these, since server-side destinations do not have a notion of `integrations`. Also we are not implementing any retry mechanism? are we?

**Review by pooyaj — COMMENTED** [review]
Couple of comments otherwise ✨

**Review by pooyaj — APPROVED** [review]
Nice 👌

File: packages/core/src/events/interfaces.ts (hunk 2 above)
Comment (chrisradek) [reply packages/core/src/events/interfaces.ts:29]: Oh, I meant why change from `type` to `interface`, but was just curious.

File: packages/core/src/events/index.ts [hunk 8]
```diff
@@ -246,6 +247,7 @@ export class EventFactory {
       messageId,
     }
 
+    validateEvent(evt)
```
Comment (chrisradek) [inline packages/core/src/events/index.ts:250]: Does doing this validation here rather than as a plugin change how errors are surfaced to the user? (I thought there was some validation already happening outside of the plugin in AnalyticsBrowser so the plugin wouldn't actually catch much anyway)

File: packages/core/src/events/index.ts [hunk 9]
```diff
@@ -28,7 +29,7 @@ export class EventFactory {
       ...this.baseEvent(),
       event,
       type: 'track',
-      properties,
+      properties: properties ?? {}, // TODO: why is this not a shallow copy like everywhere else?
```
[the remark this answers is not in the document]
Comment (silesky) [reply packages/core/src/events/index.ts:32]: The question was more like “why are we not making a shallow clones of properties here, but we do make shallow clones of other objects like traits and options in other methods. Even if potentially pointless, I was wondering if someone just forgot to apply the spread operator here  :-)

File: packages/node/src/app/analytics-node.ts [hunk 10]
```diff
@@ -7,327 +7,298 @@ import {
...
+type IdentityOptions =
   | { userId: string; anonymousId?: string }
   | { userId?: string; anonymousId: string }
 
+/** Events from CoreOptions */
+export interface NodeSegmentEventOptions {
+  context?: NodeContext
+  timestamp?: CoreOptions['timestamp']
+  anonymousId?: CoreOptions['anonymousId']
+  userId?: CoreOptions['userId']
```
Comment (chrisradek) [inline packages/node/src/app/analytics-node.ts:38]: It looks like these are in the top-level object passed to our `track et al` calls. Can we omit these in node so they don't also show up under options?

**Review by chrisradek — APPROVED** [review]
Had a comment around userId/anonymousId appearing in top-level fields and options fields, otherwise looks great!
